parla/task_runtime.py

```diff
@@ -1225,23 +1254,54 @@ def run_mapper(self):
                 # move to the mapped task scheduling.
                 break
 
-    def run_deviceq_mapper(self):
+    def _schedule_tasks(self):
+        """ Currently this doesn't do any intelligent scheduling (ordering).
+            Dequeue all ready tasks and send them to device queues in order.
+        """
+        print("[Scheduler] Schedule Phase")
         while True:
             task: Optional[TaskBase] = self._dequeue_task()
             if not task or not task.assigned:
                 logger.debug("Task %r: Failed to assign", task)
                 break
             for d in task.req.devices:
-                print("[Scheduler] Task, ", task.name, ", is enqueued onto a device Q.", sep='')
-                worker = self._worker_threads[task.req.env_no]
-                worker._enqueue_task_local(task)
+                print(f"[Scheduler] Enqueuing {task.name} to device {d}")
+                self._device_queues[d].append(task)
+
+    def _launch_tasks(self):
+        """ Iterate through free devices and launch tasks on them
+        """
+        print("[Scheduler] Launch Phase")
+        with self._monitor:
+            for dev, queue in self._device_queues.items():
+                # Make sure there's an available WorkerThread
+                if len(self._free_worker_threads) == 0:
```

Comment:
can't we move this to outside of for loop?
can avoid unnecessary instructions.

Reply:
No, because on each loop iteration we consume another worker thread. So we have to make sure there's still a free one every loop.

Perhaps if we run out of worker threads we should spawn new ones?